subprojects/plugin-development/src/main/java/org/gradle/plugin/devel/internal/precompiled/PreCompileGroovyScriptsTask.java

```diff
@@ -0,0 +1,186 @@
+    private final ClassLoaderScopeRegistry classLoaderScopeRegistry;
+    private final CompileOperationFactory compileOperationFactory;
+
+    private final Set<File> pluginSourceFiles;
+    private final List<PreCompiledScript> scriptPlugins;
+
+    private final DirectoryProperty classesDir = getProject().getObjects().directoryProperty();
+    private final DirectoryProperty metadataDir = getProject().getObjects().directoryProperty();
+
+    private final DirectoryProperty precompiledGroovyScriptsDir = getProject().getObjects().directoryProperty();
+    private final DirectoryProperty generatedPluginAdaptersDir = getProject().getObjects().directoryProperty();
+
+    @Inject
+    public PreCompileGroovyScriptsTask(ScriptCompilationHandler scriptCompilationHandler,
+                                       ClassLoaderScopeRegistry classLoaderScopeRegistry,
+                                       CompileOperationFactory compileOperationFactory,
+                                       Set<File> pluginSourceFiles,
+                                       List<PreCompiledScript> scriptPlugins) {
+        this.scriptCompilationHandler = scriptCompilationHandler;
+        this.classLoaderScopeRegistry = classLoaderScopeRegistry;
+        this.compileOperationFactory = compileOperationFactory;
+        this.pluginSourceFiles = pluginSourceFiles;
+        this.scriptPlugins = scriptPlugins;
+
+        DirectoryProperty buildDir = getProject().getLayout().getBuildDirectory();
+        this.classesDir.set(buildDir.dir("groovy-dsl/compiled-scripts/classes"));
+        this.metadataDir.set(buildDir.dir("groovy-dsl/compiled-scripts/metadata"));
+
+        this.precompiledGroovyScriptsDir.set(buildDir.dir("generated-classes/groovy-dsl-plugins/classes"));
+        this.generatedPluginAdaptersDir.set(buildDir.dir("generated-classes/groovy-dsl-plugins/java"));
+    }
+
+    @PathSensitive(PathSensitivity.RELATIVE)
+    @InputFiles
+    Set<File> getScriptFiles() {
+        return pluginSourceFiles;
+    }
+
+    @OutputDirectory
+    Provider<Directory> getPrecompiledGroovyScriptsDir() {
+        return precompiledGroovyScriptsDir;
+    }
+
+    @OutputDirectory
+    DirectoryProperty getGeneratedPluginAdaptersDir() {
+        return generatedPluginAdaptersDir;
+    }
+
+    @TaskAction
+    void compileScripts() {
+        ClassLoaderScope classLoaderScope = classLoaderScopeRegistry.getCoreAndPluginsScope().createChild("pre-compiled-scripts");
+        classLoaderScope.lock();
+
+        for (PreCompiledScript scriptPlugin : scriptPlugins) {
+            CompiledScript<? extends BasicScript, ?> pluginsBlock = compilePluginsBlock(classLoaderScope, scriptPlugin);
+            CompiledScript<? extends BasicScript, ?> buildScript = compileBuildScript(classLoaderScope, scriptPlugin);
+            generateScriptPluginAdapter(scriptPlugin, pluginsBlock, buildScript);
+        }
+
+        getProject().copy(copySpec -> {
```

Comment:
:x: please use the `FileSystemOperations` service instead so it can work with instant execution

Reply:
Nice one - way fewer tests fail with IE now. Will keep this in mind.

Reply:
The tests should have said something like "_invocation of 'Task.project' at execution time is not supported_".